Menu buttons for FormHierarchy

cooperka opened the pull request:
Implements menu buttons discussed in https://github.com/opendatakit/roadmap/issues/19

- **Add** repeat child (visible on the "repeat picker" screen)
- **Delete** repeat child (visible when viewing a repeat child)
- **Go up** (visible on any screen other than the beginning)

<img width="297" alt="screenshot with add, up" src="https://user-images.githubusercontent.com/2047062/49096574-cf037200-f238-11e8-9322-ad0a8aa097a0.png">

<img width="297" alt="screenshot with delete, up" src="https://user-images.githubusercontent.com/2047062/49096575-cf9c0880-f238-11e8-9271-ce360d383e1a.png">

Add/delete functionality already existed (swipe forward to add; long-press on question text to delete) -- this change just makes it more accessible.

#### What has been done to verify that this works as intended?

Extensive manual testing with nested-repeats-complex ([XML](https://drive.google.com/open?id=1zOvDEZz6vwQr5IBT91k7MuDMZ-hJDats) / [XLS](https://drive.google.com/open?id=1xVv682pbBMQ_MTZdNnj6aP11jhPp6fT8WQ42oEr7WaA)) (top-level "friends" and "enemies" groups; nested "friends/pets" group).

Navigating in and out of nested repeat groups, adding children, deleting children, editing names, using the back button, jumping back and forth between the FormEditor and FormHierarchy views.

#### Why is this the best possible solution? Were any other approaches considered?

Adding these buttons to the menu in FormHierarchy seems to make the most sense because that's where you get an overview of the form. FormHierarchy allows you to view children of repeat groups, and now on the same screen you can add new children. You can also delete the child currently being viewed.

Other approaches could involve showing menus elsewhere in the app, such as FormEntry, but that seems less intuitive.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

The ability to add/delete repeat instances is now easier and more obvious. The new menu buttons should make repeat group navigation more efficient. Users may not expect to see the new buttons, but no old behavior has been changed.

The behavior changes should be isolated to FormHierarchyActivity, but this activity interacts with the global state via FormController, so regressions can be avoided by ensuring that the state is not mutated in some new unexpected way.

#### Do we need any specific form for testing your changes? If so, please attach one.

Linked above.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

Not that I can find. [This page](https://docs.opendatakit.org/collect-filling-forms/#jumping-to-questions) has screenshots of FormHierarchy but no repeat groups; the new buttons would not be visible in any of the screenshots.

#### Known issues being addressed
- [x] "Go up" button displays on the first screen in some forms
- [x] Deleting the last indexed (bottom) item in a repeat group causes a jump all the way to the first page in some forms
- [ ] Repeats with `jr:count` shouldn't have buttons to add/delete
- [ ] Deleting the last remaining item in a repeat group causes a navigation into the preceding repeat group, if there are 2 in a row, rather than navigating "up" in the hierarchy

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

cooperka committed:
Create stub options menu for FormHierarchy

cooperka committed:
Add 'go up' button

cooperka committed:
Add 'add' button

cooperka committed:
Factor out static createDeleteRepeatConfirmDialog util

cooperka committed:
Add 'delete' button

cooperka committed:
Use ?iconColor for new icon fills

cooperka committed:
Never show add/delete buttons on ViewOnlyFormHierarchy

cooperka committed:
Don't show delete button while picker is active

It's confusing to show it then

cooperka committed:
Make sure 'go up' icon is always second from the edge

For consistency with the 'jump' button in FormEntry.

cooperka committed:
Make 'up' arrow thicker like the 'jump' arrow

cooperka committed:
Clean up form_hierarchy_menu formatting

cooperka committed:
Fix a few PMD rule violations

cooperka committed:
Fix unintentionally overloaded method name

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -212,8 +212,14 @@ private void updateOptionsMenu() {
         if (optionsMenu == null) return;
 
+        FormController formController = Collect.getInstance().getFormController();
+        boolean isAtBeginning = formController.isCurrentQuestionFirstInForm() && !shouldShowRepeatGroupPicker();
```
What's the condition that this is trying to express?  I'm puzzled that this is defined with the exclusion `!shouldShowRepeatGroupPicker()` and then is only used once below, in an expression that then includes `shouldShowPicker`, making said exclusion moot.

Are you intending for the Up button to be visible only for the second and subsequent repeated items in a repeat group?  If so, why?

I might have misunderstood how `isCurrentQuestionFirstInForm()` works for repeat groups, so maybe it would help if you describe the intended condition in prose first, thanks!

cooperka replied:
The logic here is cleaned up quite a bit after the commit you're commenting on, and the latest code hopefully makes more sense.

In short, you're only at the top/beginning of the hierarchy if you're both **at the first question** and **not showing the repeat picker**. The "up" button should be displayed on every screen of the hierarchy except at the beginning, because then there's nothing to go "up" to.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -212,8 +212,14 @@ private void updateOptionsMenu() {
         optionsMenu.findItem(R.id.menu_add_child).setVisible(shouldShowPicker).setEnabled(shouldShowPicker);
+
+        boolean shouldShowGoUp = shouldShowPicker || !isAtBeginning;
+        optionsMenu.findItem(R.id.menu_go_up).setVisible(shouldShowGoUp).setEnabled(shouldShowGoUp);
```
It seems like `setEnabled()` would be redundant after `setVisible()`—is there a reason to do both?

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -294,6 +302,36 @@ private String getCurrentPath() {
+        FormController formController = Collect.getInstance().getFormController();
+        FormIndex originalIndex = formController.getFormIndex();
+
+        formController.jumpToIndex(repeatIndex);
```
Could you explain in a comment what the expected relationship is between the `repeatIndex` and the result?  There isn't quite enough here for me to tell (as a reader) what the logic below is intended to do and see that it works correctly.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/DialogUtils.java:
```diff
@@ -78,6 +83,41 @@ public static void showDialog(Dialog dialog, Activity activity) {
+    /**
+     * Creates a confirm/cancel dialog for deleting repeats.
+     */
+    public static void createDeleteRepeatConfirmDialog(Context context, Runnable onDeleted, Runnable onCanceled) {
```
This method is called for its side effects rather than for its return value, so it's probably better named `showXyzDialog` instead of `createXyzDialog`.

Alternatively, it could be named `createXyzDialog`, return the dialog, and let the caller call `.show()`.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -215,6 +216,9 @@ private void updateOptionsMenu() {
         FormController formController = Collect.getInstance().getFormController();
         boolean isAtBeginning = formController.isCurrentQuestionFirstInForm() && !shouldShowRepeatGroupPicker();
 
+        boolean isInRepeat = !isAtBeginning && screenIndex != null && formController.getEvent(screenIndex) == FormEntryController.EVENT_REPEAT;
```
When can `screenIndex` ever be null?  It seems like it can't be, though it's a little hard to see (`onCreate` calls `refreshView` which calls `jumpToHierarchyStartIndex`).  There are other places in this file where `screenIndex` is assumed to be non-null without a `screenIndex != null` check; I think it would be better to assign `screenIndex = startIndex` in `onCreate` for clarity and omit the check here, than to have checks in some places and not others.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -217,13 +217,28 @@ private void updateOptionsMenu() {
+    }
+
+    /** Override to disable this button. */
+    protected void toggleDeleteButton(boolean isEnabled) {
```
The verb `toggle` doesn't seem right to me; it suggests that it changes the state of a toggle button (pressed vs. not pressed, checked vs. not checked).

`show` would be clearer (e.g. `showDeleteButton(boolean)`).  `show` is also already the word used for this purpose in other places (e.g. `shouldShowPicker`, `shouldShowGoUp`).

zestyping commented on collect_app/src/main/java/org/odk/collect/android/activities/FormHierarchyActivity.java:
```diff
@@ -217,13 +217,28 @@ private void updateOptionsMenu() {
+
+    /** Override to disable this button. */
+    protected void toggleDeleteButton(boolean isEnabled) {
+        optionsMenu.findItem(R.id.menu_delete_child).setVisible(isEnabled).setEnabled(isEnabled);
```
Also a similar question as before: why call both `setVisible` and `setEnabled`?  Wouldn't `setVisible` alone be enough?

cooperka replied:
I copied this logic from `FormEntryActivity` where menu items can be visible but disabled; however, that state isn't possible here. I'll update.

zestyping commented on collect_app/src/main/java/org/odk/collect/android/utilities/DialogUtils.java:
```diff
@@ -78,6 +83,45 @@ public static void showDialog(Dialog dialog, Activity activity) {
+        String name = formController.getLastRepeatedGroupName();
+        int repeatcount = formController.getLastRepeatedGroupRepeatCount();
+        if (repeatcount != -1) {
+            name += " (" + (repeatcount + 1) + ")";
```
In the example where the repeat section is "Friends" and each instance of the repeated group is a friend, what does this show?  Suppose there are a total of three friends (Alice, Bob, Charlie) and the currently shown friend is Bob.  Will it show "Friends (3)" because there are three friends, or "Friends (2)" because Bob is #2, or "Bob (2)"?

cooperka replied:
Friends (2):

![screenshot from 2019-01-09 12-32-21](https://user-images.githubusercontent.com/2047062/50917057-a4e1ca00-140a-11e9-8d1f-bf17d9183570.png)

But this is old logic, and I think it should be improved as you described in https://github.com/opendatakit/collect/pull/2762#issuecomment-451619959. There are probably numerous other places in the code where this type of string is built... we shoud do a quick audit if we want to update it.

zestyping reviewed (commented):
Hi—I have a few questions about this.  Some of it is due to not having a deep understanding of how repeat groups are structured and how the FormController navigates around the form—sorry!—though I think it wouldn't hurt to spell this out a little more in comments, to help others who aren't intimately familiar with this avoid breaking it by accident; the logic looks quite intricate.  I hope my comments are helpful!

In the first screenshot you posted, the individual repeated instances are shown as "Alice (1)" and "Bob (2)", and in the second, the repeated group title is "Friends (1)".  Which bits of code are responsible for formatting these labels?  I think these labels are a bit confusing as is (see also my comments on https://github.com/opendatakit/collect/pull/2762).  In particular, since "Alice (1)" and "Friends (1)" look so alike, the label doesn't make it clear whether the name identifies a repeat group title or an instance of a repeated group; I think the label should use distinct formatting to make this obvious.

If we feel good about going with my suggestion in https://github.com/opendatakit/collect/pull/2762, these labels should be formatted similarly, e.g. "1. Alice", "2. Bob", and "Friends > 1".

cooperka committed:
PR fix: Clean up menu button toggling logic

cooperka committed:
PR fix: Clarify getRepeatPromptIndex logic

cooperka committed:
PR fix: Rename showDeleteRepeatConfirmDialog

cooperka committed:
PR fix: Clean up screenIndex == null logic

Avoid setting it to null, simplify logic.

cooperka commented:
@zestyping I believe I've addressed all comments now.

I pushed 4 new commits at https://github.com/opendatakit/collect/pull/2763/commits/d73ad5b4d5a0bd932ee010934045b3bfaa40c091, and I'm now rebasing on master since it's been a while.

mmarciniak90 commented:
@cooperka I noticed that buttons are not visible in the view `g1 (1)` but are visible for view `g1 (2)` in the form [fewRepeatGroups.xml.txt](https://github.com/opendatakit/collect/files/2759107/fewRepeatGroups.xml.txt)

|![screenshot_20190115-110903](https://user-images.githubusercontent.com/3907554/51173713-4a3ae900-18b6-11e9-83cb-3a9a288b9501.png)|![screenshot_20190115-110907](https://user-images.githubusercontent.com/3907554/51173712-49a25280-18b6-11e9-949c-e33b7627fe19.png)|
|---|---|

Similar situation for form: 
[nested-indexed-repeat.xml.txt](https://github.com/opendatakit/collect/files/2759121/nested-indexed-repeat.xml.txt)

|![screenshot_20190115-111908](https://user-images.githubusercontent.com/3907554/51174170-760a9e80-18b7-11e9-9fd1-1229d1dea2d9.png)|![screenshot_20190115-111912](https://user-images.githubusercontent.com/3907554/51174180-7a36bc00-18b7-11e9-9ef7-11e28863e895.png)|
|---|---|

@cooperka please take look on this
@opendatakit-bot unlabel "needs testing"

cooperka committed:
PR fix: More accurately detect beginning of form

Previously would think it was at the beginning when it was at the first
question inside of the first group.

cooperka commented:
Thanks for the info @mmarciniak90 -- I fixed it. Previously it would think it was at the very beginning when it was at the first question inside of the first group.

@opendatakit-bot label "needs testing"

zestyping commented:
Thanks so much for all the responses and making these edits, @cooperka!  I'm sorry I didn't respond right away—I've just discovered that my Github notifications are misconfigured, so I'm fixing that and intend to be more responsive in the future.

mmarciniak90 commented:
@cooperka 
I checked form: [nested-indexed-repeat.xml.txt](https://github.com/opendatakit/collect/files/2768246/nested-indexed-repeat.xml.txt)
* If user does not add any repeat group, `Go Up` button is visible and disappears when is clicked
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51313034-e26ad680-1a4c-11e9-824d-3fadbcb96220.png">

* If repeat group is added but user opens `jump screen` from question outside the repeat group, `Go Up` button is visible and disappears when is clicked 
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/51313352-9cfad900-1a4d-11e9-9244-693894b5f63a.png">

* The same situation is for nested-repeats-complex form.

@opendatakit-bot unlabel "needs testing"

cooperka committed:
Use screenIndex instead of currIndex to detect beginning

Otherwise the index could be changed when you visit a question
then go back to the hierarchy, showing the arrow on the first
screen when it shouldn't.

cooperka committed:
Handle deleting last indexed repeat group item

Previously, deleting the last indexed item in the list
(i.e. bottom one of multiple) wouldn't trigger the
repeat picker on some forms.

But now the last remaining item (i.e. only one left) will
leave the user on a blank screen. To be fixed next.

cooperka committed:
Handle deleting last remaining repeat group item

Now that it's possible to see the repeat picker in this state,
due to the last commit.

grzesiek2010 commented:
@cooperka once you fix the bug pointed out by @mmarciniak90 please fix conflicts as well.

cooperka commented:
@mmarciniak90 that bug should be resolved now. And @grzesiek2010 it's been rebased.

There are a lot of tricky edge cases here; I tested in as many forms as I could think of and it seems to finally be working well. One remaining issue is that if there are 2 repeat groups in a row, and you delete the **last remaining item** from the second repeat, it will send you to the last index of the first repeat instead of going back a level as expected, i.e. if the structure is:

```
Top-level
  Sub-group
    Friends
      A
      B
    Enemies
      C
```

When viewing `Enemies > C` -> Click **delete** -> Sends you to `Friends > B` when it should have sent you to `Sub-group`.

This is a narrow edge case, and it's not even that much of a usability issue, but @grzesiek2010 I'd appreciate your help addressing this if you think this is a blocker to merge.

@opendatakit-bot label "needs testing"

kkrawczyk123 commented:
@cooperka, It looks like issues pointed by @mmarciniak90 have been fixed which is great! I have used the same forms as she did and I have noticed a crash when I tried to add a new nested group from the hierarchy view when used nested-indexed-repeat.xml form. I am attaching gif to help you with reproducing:
![ezgif com-video-to-gif 2](https://user-images.githubusercontent.com/33342488/51544808-f5244780-1e60-11e9-87f2-d33f62372c72.gif)
(yes I clicked on + button on the last displayed screen as it is nearly visible on the gif xd )

@opendatakit-bot unlabel "needs testing"

cooperka committed:
Fix crash when trying to get ref to end of form

cooperka commented:
I've fixed the crash itself but this case will need to be handled differently. Repeats with `jr:count` shouldn't have buttons to add/delete.

~~@grzesiek2010 do you know how we might determine if a repeat has an explicit `count`? I tried using `getCaptionPrompt().getFormElement().getAdditionalAttribute("jr", "count")` (and variations) but it always returns `null`... I'll post in Slack.~~

cooperka committed:
Check isGroupSizeLocked before showing add/delete buttons

cooperka commented:
@kkrawczyk123 forms that use specific numbers of repeat items are now supported. The add/delete buttons should **not** be visible in these types of repeat groups. Can you please check again?

@opendatakit-bot label "needs testing"

kkrawczyk123 commented:
@cooperka I am not sure if I fully understand the approach that you followed. I can see some sense in disabling deleting within the nested group when we are giving the number of the repeats by ourselves according to https://github.com/opendatakit/javarosa/pull/61 but if it's about adding the new items of the nested group I think that should be allowed and I don't understand why you disabled it. I am able to add the group when I am back to the form anyway. cc @yanokwa @grzesiek2010

grzesiek2010 commented:
I agree. The delete option is redundant in case of repeat groups with `repeat-count` number but adding new groups should be possible since we can do that on the form level.

cooperka commented:
I don't follow -- if, in the `nested-indexed-repeat` form, friend A has declared they have exactly 2 pets, are you suggesting you should be able to add a third? This seems to be disallowed, how are you able to do so? This is what I see when I'm in the form:

1. How many pets does A have?
    1. Answer `2`
    1. Swipe to next question
1. What is pet #1's name?
    1. Swipe to next question
1. What is pet #2's name?
    1. Swipe to next question
1. Add another "" group? [friend, not pet]

You can not add new pets, so I hid the "add" button for pets. There's still an "add" button for friends.

grzesiek2010 commented:
Ahhh sorry I misunderstood if we use repeat-count we can add more groups only editing that value there is no prompt to add a new group at the end so the approach is ok @kkrawczyk123 please continue testing.

kkrawczyk123 commented:
@cooperka it looks like you are right! and we were wrong. It was the only concern that I had according to that issue, so I think it is ready to go!
Verified on Androids: 4.4, 5.1, 6.0, 7.0 and 8.1.

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

grzesiek2010 commented:
@cooperka please resolve conflicts

cooperka commented:
@grzesiek2010 rebased and ready 👍
